### PR TITLE
feat(api): save delegations/revocations in batch

### PR DESCRIPTION
**Pull request by merklefruit**

### CODE REVIEW

File: crates/datastore/src/redis/redis_cache.rs [hunk 1]
```diff
@@ -543,53 +543,56 @@ impl Auctioneer for RedisCache {
...
+            // Filter out the revoked delegation.
+            let updated_delegations = delegations.retain(|delegation| {
+                signed_revocations.iter().all(|revocation| {
+                    delegation.message.delegatee_pubkey != revocation.message.delegatee_pubkey
+                })
+            });
+    
+            // Save the updated delegations back to the cache.
+            self.set(&key, &updated_delegations, None).await.map_err(AuctioneerError::RedisError)?;
+        }
```
Comment (merklefruit) [inline crates/datastore/src/redis/redis_cache.rs:593]: todo(self): double check logic

File: crates/api/src/constraints/api.rs [hunk 2]
```diff
@@ -200,42 +200,41 @@ where
...
-        let mut signed_delegation = match serde_json::from_slice::<SignedDelegation>(&body_bytes) {
-            Ok(delegation) => match delegation.message.action {
-                DELEGATION_ACTION => delegation,
-                other => {
-                    warn!(request_id = %request_id, action = other, "Invalid delegation action. expected 0");
+        // Decode the incoming request body into a `SignedDelegation` array.
+        let signed_delegations = match serde_json::from_slice::<Vec<SignedDelegation>>(&body_bytes) {
+            Ok(delegations) => {
+                let action = delegations.iter().map(|d| d.message.action).collect::<HashSet<_>>();
+                if action.len() != 1 || !action.contains(&DELEGATION_ACTION) {
```
Comment (thedevbirb) [inline crates/api/src/constraints/api.rs:207]:
```suggestion
                let are_all_actions_delegations = action.len() == 1 && action.contains(&DELEGATION_ACTION);
                if !are_all_actions_delegations {
```

Comment (thedevbirb) [reply crates/api/src/constraints/api.rs:207]: Nit but imo more readable

File: crates/api/src/constraints/api.rs [hunk 3]
```diff
@@ -275,42 +274,41 @@ where
...
-        let mut signed_revocation = match serde_json::from_slice::<SignedRevocation>(&body_bytes) {
-            Ok(revocation ) => match revocation.message.action {
-                REVOCATION_ACTION => revocation,
-                other => {
-                    warn!(request_id = %request_id, action = other, "Invalid revocation action. expected 1");
+        // Decode the incoming request body into a `SignedRevocation` array.
+        let signed_revocations = match serde_json::from_slice::<Vec<SignedRevocation>>(&body_bytes) {
+            Ok(revocations) => {
+                let action = revocations.iter().map(|r| r.message.action).collect::<HashSet<_>>();
+                if action.len() != 1 || !action.contains(&REVOCATION_ACTION) {
```
Comment (thedevbirb) [inline crates/api/src/constraints/api.rs:281]: Can you do the same here?

**Review by thedevbirb — APPROVED** [review]
Just a nit!

**Comment by merklefruit** [comment]
thanks, addressed

**Pull request merged**